File: lua/3d.lua
#!/usr/bin/env lua
local the = {bins=10, seed=10, p=2,
             file="../../moot/optimize/misc/auto93.csv"}

local BIG = 1E32
math.randomseed(the.seed)

local any,atom,cat,fmt,max,min,new,trim,split

fmt = string.format

function any(t)     return t[ math.random(#t) ] end
function atom(s)    return tonumber(s) or trim(s) end
function max(x,y)   return x >= y and x or y end
function min(x,y)   return x <= y and x or y end
function new(kl,t)  kl.__index=kl; setmetatable(t,kl); return t end
function sort(t,f)  table.sort(t,f); return t end
function trim(s)    return s:match("^%s*(.-)%s*$") end

function split(s,   t) 
  t={}; for s1 in s:gmatch("([^,]+)") do t[1+#t]=atom(s1) end; return t end

local _kat,_cat
function cat(x)
  if type(x)=="table" then return (#x==0 and _kat or _cat)(x,{}) end
  if type(x)=="number" and x//1 ~= x then return fmt("%3g",x) end 
  return tostring(x)  end 

function _kat(t,u, _show)
  _show = function(s) return not tostring(s):find"^_" end
  for k,v in pairs(t) do if _show(k) then u[1+#u]=fmt(":%s %s",k,cat(v)) end end
  return '{' .. table.concat(sort(u)," ") .. '}' end

function _cat(t,u)
  for k,v in pairs(t) do u[1+#u] = cat(v) end 
  return '{' .. table.concat(u,", ") .. '}' end


-------------------------------------------------------------------------------
local Sym,Data,SYM,DATA = {},{}

function SYM() return new(Sym, {has={}, mode=nil, most=0}) end

function Sym.add(i,x,  n)
  i.has[x] = (n or 1) + (i.has[x] or 0)
  if i.has[x] > i.most then i.most,i.mode = i.has[x],x end end

function DATA(names,    i) 
  i = new(Data,{lo={}, hi={},  names=names, _rows={}, x={}, y={}}) 
  for c,s in pairs(names) do
    if not s:find"X$" then
      if s:find"^[A-Z]" then i.lo[c], i.hi[c] = BIG, -BIG end
      (s:find"[!+-]$" and i.y or i.x)[c] = c 
      if s:find"-$" then i.y[c] = 0 end
      if s:find"+$" then i.y[c] = 1 end end end
  return i end

function Data.norm(i,c, x) 
  return x=="?" and x or (x - i.lo[c]) / (i.hi[c] - i.lo[c]+1/BIG) end

function Data.dist(i,x,y,    d,n,_dist)
  _dist = function(c,a,b)
    if a=="?" and b=="?" then return 1 
    else if not i.lo[c] then return a==b and 0 or 1 
    else a,b = i:norm(c, a), i:norm(c, b)
         a = a == "?" and a or (b > 0.5 and 0 or 1)
         b = b == "?" and b or (a > 0.5 and 0 or 1)
         return math.abs(a - b) end end end 
  d,n = 0,0
  for c in pairs(i.x) do
    n = n + 1
    d = d + _dist(c, x[c], y[c]) ^ the.p end
  return (d/n) ^ (1/the.p) end

function Data.add(i,row)
  i._rows[1 + #i._rows] = row
  for c,x in pairs(row) do
    if x ~= "?" and i.lo[c] then
      i.hi[c] = math.max(x, i.hi[c])
      i.lo[c] = math.min(x, i.lo[c]) end end end  

function main(file,   d,t)
  for line in io.lines(file) do
    if d then d:add(split(line)) else d = DATA(split(line)) end end
  return d end
 
local function poles(a)
   local far, _rows = -1, {}
   for r in pairs(D) do
      table.insert(rows, r)
      if #rows > 20 then break end
   end
   for _, r in ipairs(rows) do
      for _, r2 in ipairs(rows) do
        if r > r2 then
           local d = dist(r, r2)
           if d > far then
              far, a[1], a[2] = d, r, r2
           end
        end
      end
   end
   far = -1
   for _, r in ipairs(rows) do
      local d = dist(r, a[1]) + dist(r, a[2])
      if d > far then
        far, a[3] = d, r
      end
   end
   far = -1
   for _, r in ipairs(rows) do
      local d = dist(r, a[1]) + dist(r, a[2]) + dist(r, a[3])
      if d > far then
        a[4] = r
      end
   end
end

d=main(arg[1] or the.file)
print(cat(d))

File: lua/3d.py
#!/usr/bin/env python3
import random, sys
any = random.choice
many= random.choices

class o(dict):
    def __getattr__(i, k)   : return i[k]
    def __setattr__(i, k, v): i[k] = v

BIG, the = 1E32, o(bins= 10, seed= 10, p= 2, 
                   some=30, poles=4,
                   file= "../../moot/optimize/misc/auto93.csv")
random.seed(the.seed)

class DATA:
  def __init__(i,src):
    src = iter(src)
    i.names, i.nums, i.x,i.y, i._rows = next(src),{},{},{}, []
    for c, s in enumerate(i.names):
      if not s[-1] == "X":
        if s[0].isupper(): i.nums[c] = o(lo=BIG, hi=-BIG)
        if s[-1] == "-": i.y[c] = 0
        elif s[-1] == "+": i.y[c] = 1
        else: i.x[c]=1
    [i.add(row) for row in src]
 
  def add(i,row):
    i._rows += [row]
    for c,x in enumerate(row):
      if x != "?" and c in i.nums:
        n = i.nums[c]
        n.lo = min(x, n.lo)
        n.hi = max(x, n.hi)
  
  def norm(i,c,x): 
    n= i.nums[c]; return  (x - n.lo) / (n.hi - n.lo + 1/BIG)

  def dist(i,a,b):
    def _dist(c,x,y):
      if x == "?" and y == "?": return 1
      if c in i.nums: 
        x = i.norm(c,x) if x != "?" else (0 if y > 0.5 else 1)
        y = i.norm(c,y) if y != "?" else (0 if x > 0.5 else 1)
        return abs(x - y)
      return x != y
    lst= [ _dist(c, a[c], b[c]) ** the.p for c in i.x] 
    return (sum(lst) / len(lst)) ** (1 / the.p)

  def poles(i):
    r0, *some = many(i._rows, k=the.some+1)
    out = [max(some, key = lambda r1: i.dist(r1,r0))]
    for _ in range(the.poles-1):
      out += [max(some, key=lambda r2: sum(i.dist(r2,r1) for r1 in out))]
    return out

#----------------------------------------------------------------------------------------
def cat(x,   isa=isinstance):
  if isa(x, list)          : return "{" + ", ".join(map(cat, x)) + "}"
  if isa(x, (float,int))   : return str(x) if x ==int(x) else f"{x:.3g}"
  if isa(x, dict): return cat([f":{k} {cat(v)}" for k,v in x.items() if str(k)[0]!="_"])
  if hasattr(x,"__dict__") : return x.__class__.__name__ + cat(x.__dict__)
  return str(x)

def cast(x):
  for what in (int, float):
    try: return what(x)
    except: pass
  x = x.strip()
  y = x.lower()
  return (y== 'true') if y in ("true", "false") else x

def csv(path):
  with open(path) as f:
    for line in f:
      yield [cast(x) for x in line.strip().split(",")]

#---------------------------------------------------------------------------------------/
if __name__ == "__main__":
  f = sys.argv[1] if len(sys.argv) > 1 else the.file
  d = DATA(csv(f))
  print(cat(d))
  [print(r) for r in d.poles()]

